### ISSUE TITLE
Make some dependencies optional?

### ISSUE THREAD

**Issue by Merivuokko** [jgm/pandoc#8285]
I wonder if it would be possible to make some of Pandoc's dependencies optional.

The easiest one would probably be making it possible to disable the web server component with a cabal flag.
Disabling Lua support would also be nice, but probably more complicated.

Many Haskell libraries and applications use Pandoc API to convert documents from one format to another, and for these pieces of software, a web server is probably quite a bit of overkill. Code bloat also comes with additional problems of increased compilation times.

Personally I have a project where I translate Pandoc-flavour of Markdown to HTML 5, and would enjoy the possibility to drop off some Pandoc features.

What do you think?

**Comment by jgm** [comment]
I think there's a general policy that a library's API should not depend on cabal flags.  So that speaks against this.

In an ideal world, pandoc would be broken into a number of modular components.  (See #6215 for a start on this.)

And pandoc-server would be a separate package entirely.

The present setup, though, is much easier for developing and packaging.  For example, there's no separate pandoc-server package to distribute (you just have to rename or symlink the pandoc executable).  That makes the binary packages we distribute smaller and simpler.

Adding the server actually doesn't add too much to pandoc's transitive dependencies.  Just:  `servant`, `servant-server`, `wai`, `wai-app-static`, `wai-extra`, and `warp`.  All of the dependencies of these packages are already dependencies of pandoc.

One possibility would be to create a separate `pandoc-cli` package for the executable and a separate `pandoc-server` package for the server library; `pandoc-cli` could then depend on both `pandoc` and `pandoc-server`.
@tarleb any thoughts on that?

**Comment by jgm** [comment]
@tarleb could also comment on how hard it would be to split the Lua stuff into a separate `pandoc-lua` package.

**Comment by jgm** [comment]
Here's one possible concrete plan for splitting the pandoc package:

- `pandoc` the main library
- `pandoc-server` - Text.Pandoc.Server (depends on `pandoc`)
- `pandoc-lua` - Text.Pandoc.Lua.* (depends on `pandoc`)
- `pandoc-app` - Text.Pandoc.App and the executable (depends on `pandoc`, `pandoc-server`, and `pandoc-lua`)

**Comment by jgm** [comment]
Thinking more, one difficulty comes from the test suite. Some of the tests (Tests.Old and Tests.Command) presuppose that we have Text.Pandoc.App.  (But they don't actually require the pandoc CLI app to be built; instead of running the actual `pandoc`, they run the test suite with an `--emulate` option.)

So this might argue for:

- `pandoc` the main library, including Text.Pandoc.App and the test suite (note: only one of the current command tests uses a lua filter), 
- `pandoc-lua` - Text.Pandoc.Lua.*, Text.Pandoc.Writers.Custom, Text.Pandoc.Readers.Custom - depends on `pandoc`
- `pandoc-server` - Text.Pandoc.App - depends on `pandoc`
- `pandoc-app` or `pandoc-cli` - just the executable, depends on all three libraries above

Maybe this is a better approach.

**Comment by jgm** [comment]
This plan would remove the following dependencies from the core `pandoc` library:
```
hslua
hslua-aeson
hslua-classes
hslua-core
hslua-marshalling
hslua-module-doclayout
hslua-module-path
hslua-module-system
hslua-module-text
hslua-module-version
hslua-objectorientation
hslua-packaging
lpeg
lua
pandoc-lua-marshal
servant
servant-server
wai
wai-app-static
wai-extra
warp
unix
```

**Comment by tarleb** [comment]
I'm in favor of doing this. :+1:

**Comment by tarleb** [comment]
The biggest challenge is probably to cleanly separate the custom reader and writer from the library. I'll have to take a closer look.

**Comment by jgm** [comment]
I've started a branch where I'm giving this a try.
split-package

**Comment by jgm** [comment]
You're right that the custom writer/reader are going to be tricky, unless we were to use the first approach above (which puts T.P.App in pandoc-app and lets it depend on pandoc-lua).

**Comment by jgm** [comment]
So, we could try using the first approach, with `pandoc-app` containing T.P.App + the executable.
In that case we'd have to put the "Tests.Old" and "Tests.Command" tests in `pandoc-app` and leave the other tests in `pandoc`, which isn't ideal, but maybe is okay.

**Comment by jgm** [comment]
I guess a third approach would be to parameterize `convertWithOpts` with the Lua subsystem.

**Comment by jgm** [comment]
Fourth approach:  put a `lua` flag on `pandoc`.  The `pandoc` API is the same whether the flag is used or not. If the flag is used, then `pandoc` depends on `pandoc-lua` and custom writers and Lua filters work.  If it is not used, then functions like `writeCustom` or attempts to run Lua filters just produce errors indicating that "pandoc was not compiled with Lua support."

This approach would allow all of the testing infrastructure to stay with the pandoc library.

**Comment by ulysses4ever** [comment]
Thanks for bringing up this issue. I too wondered not once about whether it'd be possible to make pandoc more modular. I wanted to share one piece of data that you can get with the recent [dr-cabal](https://github.com/chshersh/dr-cabal) tool, which helps to assess contribution of individual dependencies into overall `cabal build all` time. So far, it only fully renders in a terminal, but this [Gist still gives all the important data](https://gist.github.com/ulysses4ever/bf15a0145f064511781cd35e1e049a4f) (you could get more specifics on where the time for each dependency is spent if you run dr-cabal on the attached file but it's fairly non-insightful).
[dr-cabal-debug.zip](https://github.com/jgm/pandoc/files/9541634/dr-cabal-debug.zip)

In a nutshell, none of the packages listed above are high in the list of slowest-to-build dependencies. But it's conceivable to expect a ~5-10% speed-up (although, this is a bit hard to calculate because of parallelism in the build process).

**Comment by jgm** [comment]
I've split off pandoc-cli and pandoc-server.

Now, as for the Lua issue, that's more complex.  It might be nice to at least be able to disable Lua support under a flag.  This would not necessarily require removing Text.Pandoc.Lua as an exported module; instead, we could simply provide versions of `applyFilter`, `readCustom`, and `writeCustom` that issue an error indicating that this version of pandoc is compiled without Lua support.  The difficulty is with the lower-level functions that Text.Pandoc.Lua currently exports:  in particular, `setGlobals` and `runLua`, both of which have types that involve `LuaE` from hslua.  And these are used by the test suite, so they can't be hidden.

My question (@tarleb) is whether they really need to have these types.  For example, `runLua` looks like this:
```hs
  runLua :: (PandocMonad m, MonadIO m)
         => LuaE PandocError a -> m (Either PandocError a)
  runLua action =
    runPandocLua . try $ do
      initLuaState
      liftPandocLua action
```
where `liftPandocLua` is just the `PandocLua` constructor.
Why not, then, simply redefine `runLua` as
```hs
  runLua :: (PandocMonad m, MonadIO m)
         => PandocLua a -> m (Either PandocError a)
  runLua action = try 
    runPandocLua . try $ do
      initLuaState
      action
```
We could export this without needing to export (or import) `LuaE`.  And we could provide a dummy version that raises an error when compiled without the lua flag.

(Side question:  are both `try`s really needed?)

**Comment by tarleb** [comment]
I'm not sure tbh, but I think that should work.

Maybe we could parameterize all "custom" (Lua) functions over a new type

``` haskell
data ScriptingEngine = ScriptingEngine
  { applyFilter :: (PandocMonad m, MonadIO m) => FilePath -> Pandoc -> m Pandoc
  , readCustom :: ...
  , writeCustom :: ...
  }
```

Then, if somebody has the time to swap Lua with guile, python, or duktape, they'd have a clear path to do so.

**Comment by jgm** [comment]
OK, I tried this and it doesn't quite work, because the test suite wants to pass a bunch of low-level Lua stuff to runLua. So we'd need to export `liftPandocLua`, and its type signature includes `LuaE`.

**Comment by jgm** [comment]
Here's the patch if you're interested:
``` diff
diff --git a/src/Text/Pandoc/Lua.hs b/src/Text/Pandoc/Lua.hs
index b1d263fc7..eb0ecba95 100644
--- a/src/Text/Pandoc/Lua.hs
+++ b/src/Text/Pandoc/Lua.hs
@@ -20,7 +20,8 @@ module Text.Pandoc.Lua
   ) where
 
 import Text.Pandoc.Lua.Filter (applyFilter)
-import Text.Pandoc.Lua.Global (Global (..), setGlobals)
+import Text.Pandoc.Lua.Global (Global (..))
+import Text.Pandoc.Lua.PandocLua (setGlobals)
 import Text.Pandoc.Lua.Init (runLua)
 import Text.Pandoc.Lua.Reader (readCustom)
 import Text.Pandoc.Lua.Writer (writeCustom)
diff --git a/src/Text/Pandoc/Lua/Filter.hs b/src/Text/Pandoc/Lua/Filter.hs
index b920f358d..86bd414e9 100644
--- a/src/Text/Pandoc/Lua/Filter.hs
+++ b/src/Text/Pandoc/Lua/Filter.hs
@@ -21,7 +21,8 @@ import Text.Pandoc.Definition
 import Text.Pandoc.Lua.ErrorConversion ()
 import Text.Pandoc.Lua.Marshal.AST
 import Text.Pandoc.Lua.Marshal.Filter
-import Text.Pandoc.Lua.Global (Global (..), setGlobals)
+import Text.Pandoc.Lua.Global (Global (..))
+import Text.Pandoc.Lua.PandocLua (setGlobals, liftPandocLua)
 import Text.Pandoc.Lua.Init (runLua)
 import Control.Exception (throw)
 import qualified Data.Text as T
@@ -63,7 +64,7 @@ applyFilter :: (PandocMonad m, MonadIO m)
 applyFilter globals fp doc = do
   runLua >=> forceResult fp $ do
     setGlobals globals
-    runFilterFile fp doc
+    liftPandocLua $ runFilterFile fp doc
 
 forceResult :: (PandocMonad m, MonadIO m)
             => FilePath -> Either PandocError Pandoc -> m Pandoc
diff --git a/src/Text/Pandoc/Lua/Global.hs b/src/Text/Pandoc/Lua/Global.hs
index 938cc84b7..495cf6393 100644
--- a/src/Text/Pandoc/Lua/Global.hs
+++ b/src/Text/Pandoc/Lua/Global.hs
@@ -11,7 +11,7 @@ Pandoc's Lua globals.
 -}
 module Text.Pandoc.Lua.Global
   ( Global (..)
-  , setGlobals
+  , setGlobal
   ) where
 
 import HsLua as Lua
@@ -41,10 +41,6 @@ data Global =
   | PANDOC_VERSION
   -- Cannot derive instance of Data because of CommonState
 
--- | Set all given globals.
-setGlobals :: [Global] -> LuaE PandocError ()
-setGlobals = mapM_ setGlobal
-
 setGlobal :: Global -> LuaE PandocError ()
 setGlobal global = case global of
   -- This could be simplified if Global was an instance of Data.
diff --git a/src/Text/Pandoc/Lua/Init.hs b/src/Text/Pandoc/Lua/Init.hs
index 966f0a581..e02a7c793 100644
--- a/src/Text/Pandoc/Lua/Init.hs
+++ b/src/Text/Pandoc/Lua/Init.hs
@@ -40,11 +40,11 @@ import qualified Text.Pandoc.Lua.Module.Utils as Pandoc.Utils
 -- | Run the lua interpreter, using pandoc's default way of environment
 -- initialization.
 runLua :: (PandocMonad m, MonadIO m)
-       => LuaE PandocError a -> m (Either PandocError a)
+       => PandocLua a -> m (Either PandocError a)
 runLua action =
   runPandocLua . try $ do
     initLuaState
-    liftPandocLua action
+    action
 
 -- | Modules that are loaded at startup and assigned to fields in the
 -- pandoc module.
diff --git a/src/Text/Pandoc/Lua/PandocLua.hs b/src/Text/Pandoc/Lua/PandocLua.hs
index 52ace5f6b..b4d87f3a7 100644
--- a/src/Text/Pandoc/Lua/PandocLua.hs
+++ b/src/Text/Pandoc/Lua/PandocLua.hs
@@ -22,6 +22,7 @@ module Text.Pandoc.Lua.PandocLua
   ( PandocLua (..)
   , runPandocLua
   , liftPandocLua
+  , setGlobals
   ) where
 
 import Control.Monad.Catch (MonadCatch, MonadMask, MonadThrow)
@@ -30,7 +31,7 @@ import Control.Monad.IO.Class (MonadIO)
 import HsLua as Lua
 import Text.Pandoc.Class.PandocMonad (PandocMonad (..))
 import Text.Pandoc.Error (PandocError)
-import Text.Pandoc.Lua.Global (Global (..), setGlobals)
+import Text.Pandoc.Lua.Global (Global (..), setGlobal)
 import Text.Pandoc.Lua.Marshal.CommonState (peekCommonState)
 
 import qualified Control.Monad.Catch as Catch
@@ -52,6 +53,10 @@ newtype PandocLua a = PandocLua { unPandocLua :: LuaE PandocError a }
 liftPandocLua :: LuaE PandocError a -> PandocLua a
 liftPandocLua = PandocLua
 
+-- | Set all given globals.
+setGlobals :: [Global] -> PandocLua ()
+setGlobals = liftPandocLua . mapM_ setGlobal
+
 -- | Evaluate a @'PandocLua'@ computation, running all contained Lua
 -- operations..
 runPandocLua :: (PandocMonad m, MonadIO m) => PandocLua a -> m a
@@ -60,7 +65,7 @@ runPandocLua pLua = do
   globals <- defaultGlobals
   (result, newState) <- liftIO . Lua.run . unPandocLua $ do
     putCommonState origState
-    liftPandocLua $ setGlobals globals
+    setGlobals globals
     r <- pLua
     c <- getCommonState
     return (r, c)
@@ -108,6 +113,6 @@ instance PandocMonad PandocLua where
   getCommonState = PandocLua $ do
     Lua.getglobal "PANDOC_STATE"
     forcePeek $ peekCommonState Lua.top
-  putCommonState = PandocLua . setGlobals . (:[]) . PANDOC_STATE
+  putCommonState = setGlobals . (:[]) . PANDOC_STATE
 
   logOutput = IO.logOutput
diff --git a/src/Text/Pandoc/Lua/Reader.hs b/src/Text/Pandoc/Lua/Reader.hs
index 6303dace3..27fdfc725 100644
--- a/src/Text/Pandoc/Lua/Reader.hs
+++ b/src/Text/Pandoc/Lua/Reader.hs
@@ -20,7 +20,7 @@ import HsLua as Lua hiding (Operation (Div))
 import Text.Pandoc.Definition
 import Text.Pandoc.Class (PandocMonad, findFileWithDataFallback, report)
 import Text.Pandoc.Logging
-import Text.Pandoc.Lua.Global (Global (..), setGlobals)
+import Text.Pandoc.Lua.Global (Global (..))
 import Text.Pandoc.Lua.Init (runLua)
 import Text.Pandoc.Lua.PandocLua
 import Text.Pandoc.Lua.Marshal.Pandoc (peekPandoc)
@@ -36,12 +36,13 @@ readCustom luaFile opts srcs = do
   luaFile' <- fromMaybe luaFile <$> findFileWithDataFallback "readers" luaFile
   res <- runLua $ do
     setGlobals globals
-    stat <- dofileTrace luaFile'
-    -- check for error in lua script (later we'll change the return type
-    -- to handle this more gracefully):
-    when (stat /= Lua.OK)
-      Lua.throwErrorAsException
-    parseCustom
+    liftPandocLua $ do
+      stat <- dofileTrace luaFile'
+      -- check for error in lua script (later we'll change the return type
+      -- to handle this more gracefully):
+      when (stat /= Lua.OK)
+        Lua.throwErrorAsException
+      parseCustom
   case res of
     Left msg -> throw msg
     Right doc -> return doc
diff --git a/src/Text/Pandoc/Lua/Writer.hs b/src/Text/Pandoc/Lua/Writer.hs
index 6b4273d88..d7f9e4630 100644
--- a/src/Text/Pandoc/Lua/Writer.hs
+++ b/src/Text/Pandoc/Lua/Writer.hs
@@ -23,7 +23,8 @@ import HsLua
 import Control.Monad.IO.Class (MonadIO)
 import Text.Pandoc.Class (PandocMonad, findFileWithDataFallback)
 import Text.Pandoc.Definition (Pandoc (..))
-import Text.Pandoc.Lua.Global (Global (..), setGlobals)
+import Text.Pandoc.Lua.Global (Global (..))
+import Text.Pandoc.Lua.PandocLua (setGlobals, liftPandocLua)
 import Text.Pandoc.Lua.Init (runLua)
 import Text.Pandoc.Options (WriterOptions)
 import qualified Text.Pandoc.Lua.Writer.Classic as Classic
@@ -38,7 +39,7 @@ writeCustom luaFile opts doc = do
                , PANDOC_SCRIPT_FILE luaFile'
                , PANDOC_WRITER_OPTIONS opts
                ]
-    dofileTrace luaFile' >>= \case
+    liftPandocLua $ dofileTrace luaFile' >>= \case
       OK -> pure ()
       _  -> throwErrorAsException
     -- Most classic writers contain code that throws an error if a global
@@ -50,7 +51,7 @@ writeCustom luaFile opts doc = do
           pushName x
           rawget (nth 2) <* remove (nth 2) -- remove global table
 
-    rawgetglobal "Writer" >>= \case
+    liftPandocLua $ rawgetglobal "Writer" >>= \case
       TypeNil -> do
         pop 1  -- remove nil
         Classic.runCustom opts doc
```

**Comment by tarleb** [comment]
The package has been split into *pandoc*, *pandoc-lua*, *pandoc-server*, and *pandoc-cli*. The new pandoc-cli package allows to remove Lua and the server from the binary via cabal flags. Any finer grained packages are not planned at the moment; see #6215 and #8340 for relevant discussions.

**Comment by jgm** [comment]
@tarleb I've added a tool that can help us untangle module dependencies and figure out whether any further splitting would be useful.
Example:
``` 
make moduledeps ROOTNODE=T.P.Parsing
T.P.Parsing
 T.P.Sources
 T.P.Parsing.Capabilities
  T.P.Logging
   T.P.Shared
    T.P.Extensions
    T.P.Asciify
  T.P.Options
   T.P.UTF8
   T.P.Highlighting
    T.P.Class
     T.P.Class.PandocPure
      T.P.Error
      T.P.Class.CommonState
       T.P.MediaBag
        T.P.MIME
       T.P.Translations
      T.P.Class.PandocMonad
       T.P.Data
     T.P.Class.PandocIO
      T.P.Class.IO
     T.P.Class.Sandbox
  T.P.Parsing.Types
  T.P.Readers.LaTeX.Types
 T.P.Parsing.Citations
  T.P.Parsing.General
   T.P.Parsing.State
   T.P.XML
 T.P.Parsing.GridTable
 T.P.Parsing.Lists
 T.P.Parsing.Math
 T.P.Parsing.Smart
```

**Comment by tarleb** [comment]
Oh wow! Very helpful indeed!
